lile.go

```diff
@@ -72,43 +146,163 @@ func Server(r registerImplementation) {
+		if rcErr == nil {
+			err := rc.DeRegister(service.Name)
+			if err != nil {
+				logrus.Errorf("Failed to deregister service by id: '%s'. Error: %v", service.ID, err)
+			} else {
+				logrus.Infof("Deregistered service '%s' at consul.", service.ID)
+			}
+		}
+	}
+
+	grpcServer.GracefulStop()
+	ctx, cancel := context.WithTimeout(context.TODO(), 5 *time.Second)
+	defer cancel()
+	if err := prometheusServer.Shutdown(ctx); err != nil {
+		logrus.Fatalf("Timed out during shutdown of prometheus server. Error: %v", err)
+	}
+
+	if service.Config.UsePubSub {
+		logrus.Infof("Unsubscribing...")
+		// TODO: unsubscribe pubsub.Unsubscribe(service.Subscriber)
+		// right now signal takes care of stopping the pubsub.Subscribe goroutine
+	}
+}
+
+func Run() error {
+	var err error
+	serviceListener, err = net.Listen("tcp", service.Address())
 	if err != nil {
 		return err
 	}
 
-	logrus.Infof("Serving gRPC on %s", service.Port)
-	return CreateServer().Serve(lis)
+	// add to registry
+	if service.Config.UseConsul {
+		rc, rcErr := getRegistryClient()
+		if rcErr == nil {
+			err := rc.Register(service.ID, service.Name, service.Config.Service.Port, &api.AgentServiceCheck{
+				CheckID:                        "",
+				Name:                           "",
+				Args:                           nil,
+				Script:                         "",
+				DockerContainerID:              "",
+				Shell:                          "",
+				Interval:                       "",
+				Timeout:                        "",
+				TTL:                            "",
+				HTTP:                           "",
+				Header:                         nil,
+				Method:                         "",
+				TCP:                            "",
+				Status:                         "",
+				Notes:                          "",
+				TLSSkipVerify:                  false,
+				GRPC:                           "",
+				GRPCUseTLS:                     false,
+				DeregisterCriticalServiceAfter: "",
+			})
+
+			if err != nil {
+				logrus.Errorf("Failed to register service at '%s'. error: %v", service.Config.RegistryAddress, err)
+			} else {
+				logrus.Infof("Regsitered service '%s' at consul.", service.ID)
+			}
+		} else {
+			logrus.Errorf("Failed to create registry client with address: '%s'. Error: %v", service.Config.RegistryAddress, err)
+		}
+	}
+
+	// setup pubsub
+	if service.Config.UsePubSub {
+		logrus.Infof("Subscribing...")
+		// TODO: make Subscribe return without waiting for signal
+		go pubsub.Subscribe(service.Subscriber)
+	}
+
+	logrus.Infof("Serving gRPC on %s", service.Address())
+	return CreateServer().Serve(serviceListener)
+}
+
+func getRegistryClient() (registry.Client, error) {
+	rc, err := registry.NewConsulClient(service.Config.RegistryAddress)
+	if err != nil {
+		logrus.Errorf("Failed to create wuth registry address '%s'", service.Config.RegistryAddress)
+		return nil, err
+	}
+
+	return rc, nil
 }
 
 func CreateServer() *grpc.Server {
+	zap, err := zap.NewProduction()
```

Comment:
I think we should probably leave this out, we found logging to be far too verbose for production (we use metrics instead), I am actually thinking that would extract all this into a `DefaultMiddleware` option and then users can opt out and do whatever

Reply:
Also because it adds another logger and we have logrus, which should be abstracted away too

Reply:
Removed